File: Minimum Multiplications to reach End[07-Septeber-2023]/Solution.cpp
//  >> CPP CODE <<

class Solution {
  public:
    int minimumMultiplications(vector<int>& arr, int start, int end) {
        
        int level = 0;
        
        bool visited[100001];
        memset(visited, false, sizeof(visited));
    
        queue<int> q;
        q.push(start);
        visited[start] = true;
    
        while (!q.empty()) {
            int size = q.size();
        
            for (int i = 0; i < size; i++) {
                int current = q.front();
                q.pop();
                
                if (current == end) {
                    return level;
                }
            
                for (int number : arr) {
                    int newStart = (current * number) % 100000;
                    if (!visited[newStart]) {
                        q.push(newStart);
                        visited[newStart] = true;
                    }
                }
            }
        
            level++;
        }

            return -1;
    }
};

File: Minimum Multiplications to reach End[07-Septeber-2023]/Solution.py
#  >> PYTHON CODE <<

from typing import List

from collections import deque

class Solution:
    def minimumMultiplications(self, arr, start, end):
        level = 0
        queue = deque()
        visited = [False] * 100001
        
        queue.append(start)
        visited[start] = True
        
        while queue:
            size = len(queue)
            
            for i in range(size):
                current = queue.popleft()
                
                if current == end:
                    return level
                
                for number in arr:
                    newStart = (current * number) % 100000
                    if not visited[newStart]:
                        queue.append(newStart)
                        visited[newStart] = True
            
            level += 1
        
        return -1